README.md

```diff
@@ -1,2 +1,208 @@
+```
+
+__Pros__
+
+* New common make features will be pulled in automatically as they are added (although see the caveat in the __Cons__ section)
+* Depending on the download recipe used, package maintainers have flexibility in whether they will automatically pick up none, some, or all updates
+
+__Cons__
+
+* All users, including the package builder, must have network access in order to build the package, at least initially
+* Developers may need to periodically to refresh the common make files if another developer starts using the features of a newer version
+
+#### Local copy
+
+In this option, a copy of the appropriate `*-common.mk` file(s) are downloaded manually and
+checked into the package being built.  In this case the Makefile can just use the
+`include` directive with no additional complications.
+
+__Pros__
+
+* The Makefile only needs a simple include statement
+* The package is self-contained and fully in control of pulling in future common make infrastructure updates
+
+__Cons__
+
+* Developers may make local changes to the common make infrastructure which are not fed upstream, making future updates painful or even infeasible in the worst cases
+* The package will not pick up any common make updates without manual intervention
+
+
+### Overriding variables
+
+Overriding variables should be done before referencing the common make recipes.  E.g., to
+override `GOFLAGS` so that package names are printed as they are compiled your
+`Makefile` might look like
+
+```make
+GOFLAGS=-v
+
+include go-common.mk
+```
+
+### Adding to targets
+
+All public targets in the common make recipes are so-called double colon targets.  This
+means that additional targets with the same name can be defined and they will be run in
+addition to the common make targets.
+
+These additional targets must also be double colon targets, since Make does not allow a
+target to have both single and double colon versions.
+
+As trivial example, to print a message that the main build step was about to start your
+`Makefile` might look like this:
+
+```make
+pre-build::
+	@echo "Main build about to start"
+
+include go-common.mk
+```
+
+Note that the `pre-` and `post-` targets are all designed to specifically allow for
+this and are empty in the common make infrastructure.
+
+### Overriding targets
+
+There may be instances where you want to use some targets from the common make recipes
+but want to override others.  This requires a little more work, but can be done.
+
+Before doing this, it is worth asking whether what you want to achieve should be
+supported enhancement.  If so, please consider making such a change in preference from using
+this technique.
+
+To completely override the `build` target, for example, so that none of the common make
+`build` target is run, but all of the other common make targets are available, your Makefile might look like this:
+
+```make
+build:
+	go build ./...
+
+%: force
```

Comment:
How does this work? It matches any command and attempts to use them from go-common.mk if they aren't defined above? Not sure what part force plays in this.

Reply:
> How does this work? It matches any command and attempts to use them from go-common.mk if they aren't defined above? Not sure what part force plays in this.

This is best explained in https://www.gnu.org/software/make/manual/html_node/Overriding-Makefiles.html